### PR TITLE
Upgrade tinkerpop version to 3.4.3

### PR DESCRIPTION
**Pull request by Linary**
Change-Id: I5443b6a05eea74205a373f093dad27536e04f0be

### CODE REVIEW

File: hugegraph-api/src/main/java/com/baidu/hugegraph/auth/ContextGremlinServer.java [hunk 1]
```diff
@@ -40,27 +38,21 @@
...
     public ContextGremlinServer(final Settings settings) {
         /*
-         * TODO: use GremlinServer(Settings, ExecutorService)
          * which can be obtained from https://github.com/apache/tinkerpop/pull/813
          *
          * NOTE: should call GremlinServer::configureMetrics() but it's private
          * settings.optionalMetrics().ifPresent(GremlinServer::configureMetrics)
          */
-        super(create(settings));
+        super(settings, newGremlinExecutorService(settings));
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/auth/ContextGremlinServer.java]: update more comments

File: hugegraph-core/src/main/java/com/baidu/hugegraph/plugin/HugeGraphGremlinPlugin.java [hunk 2]
```diff
@@ -19,38 +19,51 @@
...
+        imports = DefaultImportCustomizer.build()
+                                         .addClassImports(classes)
+                                         .create();
     }
 
-    @Override
-    public void pluginTo(final PluginAcceptor pluginAcceptor) {
-        pluginAcceptor.addImports(IMPORTS);
+    private static final HugeGraphGremlinPlugin instance =
+            new HugeGraphGremlinPlugin();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/plugin/HugeGraphGremlinPlugin.java]: also assign in static block

File: hugegraph-dist/src/assembly/static/conf/gremlin-server.yaml [hunk 3]
```diff
@@ -5,13 +5,14 @@ channelizer: org.apache.tinkerpop.gremlin.server.channel.HttpChannelizer
...
   gremlin-groovy: {
-    imports: [java.lang.Math],
-    staticImports: [java.lang.Math.PI],
-    scripts: [scripts/empty-sample.groovy]
+    plugins: {
+      com.baidu.hugegraph.plugin.HugeGraphGremlinPlugin: {},
+      org.apache.tinkerpop.gremlin.server.jsr223.GremlinServerGremlinPlugin: {},
+      org.apache.tinkerpop.gremlin.jsr223.ImportGremlinPlugin: {classImports: [java.lang.Math], methodImports: [java.lang.Math#*]},
+      org.apache.tinkerpop.gremlin.jsr223.ScriptFileGremlinPlugin: {files: [scripts/empty-sample.groovy]}
+    }
```
Comment (javeme) [inline hugegraph-dist/src/assembly/static/conf/gremlin-server.yaml]: import DateUtil

**Comment by Linary** [comment]
```
public void shouldConstructDetachedEdge() {
        g.E(convertToEdgeId("marko", "knows", "vadas")).next().property("year", 2002);
        final DetachedEdge detachedEdge = DetachedFactory.detach(g.E(convertToEdgeId("marko", "knows", "vadas")).next(), true);
...
```

```
com.baidu.hugegraph.backend.BackendException: Can't do index query when there are changes in transaction

	at com.baidu.hugegraph.backend.tx.GraphIndexTransaction.queryIndex(GraphIndexTransaction.java:331)
	at com.baidu.hugegraph.backend.tx.GraphTransaction.indexQuery(GraphTransaction.java:1044)
	at com.baidu.hugegraph.backend.tx.GraphTransaction.query(GraphTransaction.java:330)
	at com.baidu.hugegraph.backend.tx.GraphTransaction.queryVerticesFromBackend(GraphTransaction.java:513)
	at com.baidu.hugegraph.backend.cache.CachedGraphTransaction.queryVerticesFromBackend(CachedGraphTransaction.java:150)
	at com.baidu.hugegraph.backend.tx.GraphTransaction.queryVertices(GraphTransaction.java:482)
	at com.baidu.hugegraph.HugeGraph.vertices(HugeGraph.java:390)
	at com.baidu.hugegraph.traversal.optimize.HugeGraphStep.vertices(HugeGraphStep.java:94)
	at com.baidu.hugegraph.traversal.optimize.HugeGraphStep.lambda$new$0(HugeGraphStep.java:66)
	at org.apache.tinkerpop.gremlin.process.traversal.step.map.GraphStep.processNextStart(GraphStep.java:155)
	at org.apache.tinkerpop.gremlin.process.traversal.step.util.AbstractStep.hasNext(AbstractStep.java:143)
```

**Comment by Linary** [comment]
image

```
org.apache.tinkerpop.shaded.jackson.core.JsonGenerationException: Can not write a field name, expecting a value

	at org.apache.tinkerpop.shaded.jackson.core.JsonGenerator._reportError(JsonGenerator.java:1961)
	at org.apache.tinkerpop.shaded.jackson.core.json.UTF8JsonGenerator.writeFieldName(UTF8JsonGenerator.java:188)
	at org.apache.tinkerpop.shaded.jackson.core.json.JsonGeneratorImpl.writeStringField(JsonGeneratorImpl.java:202)
	at org.apache.tinkerpop.gremlin.structure.util.star.StarGraphGraphSONSerializerV1d0.ser(StarGraphGraphSONSerializerV1d0.java:77)
	at org.apache.tinkerpop.gremlin.structure.util.star.StarGraphGraphSONSerializerV1d0.serializeWithType(StarGraphGraphSONSerializerV1d0.java:67)
	at org.apache.tinkerpop.gremlin.structure.util.star.StarGraphGraphSONSerializerV1d0.serializeWithType(StarGraphGraphSONSerializerV1d0.java:50)
	at org.apache.tinkerpop.shaded.jackson.databind.ser.impl.TypeWrappedSerializer.serialize(TypeWrappedSerializer.java:32)
	at org.apache.tinkerpop.shaded.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
	at org.apache.tinkerpop.shaded.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319)
	at org.apache.tinkerpop.shaded.jackson.databind.ObjectMapper._configAndWriteValue(ObjectMapper.java:3906)
	at org.apache.tinkerpop.shaded.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:3177)
	at org.apache.tinkerpop.gremlin.structure.io.graphson.GraphSONWriter.writeVertex(GraphSONWriter.java:82)
	at org.apache.tinkerpop.gremlin.structure.io.IoVertexTest.shouldReadWriteVertexWithBOTHEdges(IoVertexTest.java:129)
```

**Comment by Linary** [comment]
image

File: hugegraph-api/src/main/java/com/baidu/hugegraph/auth/ContextGremlinServer.java (hunk 1 above)
Comment (Linary) [reply hugegraph-api/src/main/java/com/baidu/hugegraph/auth/ContextGremlinServer.java]: GremlinServer::configureMetrics() is still private

File: hugegraph-core/src/main/java/com/baidu/hugegraph/structure/HugeEdge.java [hunk 4]
```diff
@@ -52,15 +52,18 @@
     protected EdgeLabel label;
     protected String name;
 
-    // The Vertex who owned me
-    protected HugeVertex ownerVertex;
     protected HugeVertex sourceVertex;
     protected HugeVertex targetVertex;
+    protected boolean isOutEdge;
+    protected boolean selfLoop;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/structure/HugeEdge.java]: add a method instead

File: hugegraph-core/src/main/java/com/baidu/hugegraph/structure/HugeEdge.java [hunk 5]
```diff
@@ -421,7 +406,9 @@ public HugeVertex otherVertex(HugeVertex vertex) {
     }
 
     public HugeVertex otherVertex() {
-        return this.otherVertex(this.ownerVertex);
+        return this.sourceVertex == this.ownerVertex() ?
+               this.targetVertex :
+               this.sourceVertex;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/structure/HugeEdge.java]: return this.isOutEdge ? this.targetVertex : this.sourceVertex;

File: hugegraph-test/src/main/java/com/baidu/hugegraph/tinkerpop/TestGraph.java [hunk 6]
```diff
@@ -858,4 +857,67 @@ private void initBasicVertexLabelAndEdgeLabelExceptV(String defaultVL) {
               .by("gremlin.partitionGraphStrategy.partition")
               .ifNotExist().create();
     }
+
+    public void initDefaultKnowsEdgeLabel(String defaultVL) {
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/tinkerpop/TestGraph.java]: initEdgeLabelDefaultKnowsDefault

File: hugegraph-test/src/main/java/com/baidu/hugegraph/tinkerpop/TestGraphProvider.java [hunk 7]
```diff
@@ -389,15 +398,28 @@ public void loadGraphData(final Graph graph,
 
         switch (loadGraphWith) {
             case GRATEFUL:
+                LOG.info("Load graph with {} schema",
+                         LoadGraphWith.GraphData.GRATEFUL);
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/tinkerpop/TestGraphProvider.java]: use var loadGraphWith, and change to log.debug

Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/tinkerpop/TestGraphProvider.java]: use var loadGraphWith, and change to log.debug and move to line 398

File: hugegraph-test/src/main/resources/fast-methods.filter [hunk 8]
```diff
@@ -143,6 +143,13 @@ org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasTest.Traversals.g_
 org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasTest.Traversals.g_V_hasXage_withoutX27_29X_count: Unsupported relation 'NEQ'
 org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasTest.Traversals.g_V_hasXname_gtXmX_andXcontainingXoXXX: Unsupported predicate 'containing(o)'
 
+# Unsupport edge label 'created': 'software' -> 'person'
+org.apache.tinkerpop.gremlin.process.traversal.step.map.AddEdgeTest.Traversals.g_V_hasXname_markoX_asXaX_outEXcreatedX_asXbX_inV_addEXselectXbX_labelX_toXaX: Unsupport edge from inV to outV
```
Comment (javeme) [inline hugegraph-test/src/main/resources/fast-methods.filter]: can we make an edge label for it

Comment (Linary) [reply hugegraph-test/src/main/resources/fast-methods.filter]: seems no way, there is existed an edgelabel (person -> software) in this case

File: hugegraph-core/src/main/java/com/baidu/hugegraph/type/define/Directions.java [hunk 9]
```diff
@@ -38,6 +38,7 @@ public String string() {
 
     public HugeType type() {
         switch (this) {
+//            case BOTH:
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/type/define/Directions.java]: remove it and update unit test

File: hugegraph-api/src/main/java/com/baidu/hugegraph/auth/ContextGremlinServer.java (hunk 1 above)
Comment (javeme) [reply hugegraph-api/src/main/java/com/baidu/hugegraph/auth/ContextGremlinServer.java]: will call configureMetrics() in super class

File: hugegraph-test/src/main/resources/fast-methods.filter [hunk 10]
```diff
@@ -127,6 +127,29 @@ org.apache.tinkerpop.gremlin.process.traversal.step.ComplexTest.Traversals.class
...
+org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasTest.Traversals.g_V_hasXname_not_startingWithXmarXX: Unsupported predicate 'notStartingWith(mar)'
+org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasTest.Traversals.g_V_hasXname_not_containingXarkXX: Unsupported predicate 'notContaining(ark)'
+org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasTest.Traversals.g_V_hasXname_not_endingWithXasXX: Unsupported predicate 'notEndingWith(as)'
+org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasTest.Traversals.g_V_hasXage_withoutX27_29X_count: Unsupported relation 'NEQ'
+org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasTest.Traversals.g_V_hasXname_gtXmX_andXcontainingXoXXX: Unsupported predicate 'containing(o)'
+
+# Unsupport edge label 'created': 'software' -> 'person'
+org.apache.tinkerpop.gremlin.process.traversal.step.map.AddEdgeTest.Traversals.g_V_hasXname_markoX_asXaX_outEXcreatedX_asXbX_inV_addEXselectXbX_labelX_toXaX: Unsupport edge from inV to outV
+
+# Can't customize vertex id when id strategy is 'AUTOMATIC' for vertex label 'person'
```
Comment (javeme) [inline hugegraph-test/src/main/resources/fast-methods.filter]: update more clearly

File: hugegraph-test/src/main/resources/fast-methods.filter [hunk 11]
```diff
@@ -127,6 +127,29 @@ org.apache.tinkerpop.gremlin.process.traversal.step.ComplexTest.Traversals.class
 org.apache.tinkerpop.gremlin.process.traversal.step.ComplexTest.Traversals.playlistPaths: long time
 org.apache.tinkerpop.gremlin.process.traversal.step.sideEffect.TreeTest.Traversals.g_V_out_out_treeXaX_capXaX: long time
 
+# unsupported predicate
```
Comment (javeme) [inline hugegraph-test/src/main/resources/fast-methods.filter]: update more clearly

File: hugegraph-test/pom.xml [hunk 12]
```diff
@@ -145,30 +145,6 @@
...
-                        <configuration>
-                            <testSourceDirectory>${basedir}/src/main/java/
-                            </testSourceDirectory>
-                            <testClassesDirectory>${basedir}/target/classes/
-                            </testClassesDirectory>
-                            <includes>
-                                <include>**/ProcessPerformanceTest.java</include>
-                            </includes>
-                        </configuration>
-                    </execution>
```
Comment (javeme) [inline hugegraph-test/pom.xml]: update jenkins and travis reference

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeGraphStep.java [hunk 13]
```diff
@@ -73,7 +73,11 @@ public HugeGraphStep(final GraphStep<S, E> originGraphStep) {
         LOG.debug("HugeGraphStep.vertices(): {}", this);
 
         HugeGraph graph = (HugeGraph) this.getTraversal().getGraph().get();
-        if (this.ids != null && this.ids.length > 0) {
+        // g.V().hasId(EMPTY_LIST) will set ids to null
+        if (this.ids == null) {
+            return Collections.emptyIterator();
+        }
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeGraphStep.java]: add test: query by page and return none result

File: hugegraph-test/src/main/resources/fast-methods.filter [hunk 14]
```diff
@@ -143,10 +142,10 @@ org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasTest.Traversals.g_
 org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasTest.Traversals.g_V_hasXage_withoutX27_29X_count: Unsupported relation 'NEQ'
 org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasTest.Traversals.g_V_hasXname_gtXmX_andXcontainingXoXXX: Unsupported predicate 'containing(o)'
 
-# Unsupport edge label 'created': 'software' -> 'person'
+# Unsupport edge label 'created': 'software' -> 'person', has existed an edgelabel (created: person -> software) in this case
 org.apache.tinkerpop.gremlin.process.traversal.step.map.AddEdgeTest.Traversals.g_V_hasXname_markoX_asXaX_outEXcreatedX_asXbX_inV_addEXselectXbX_labelX_toXaX: Unsupport edge from inV to outV
 
-# Can't customize vertex id when id strategy is 'AUTOMATIC' for vertex label 'person'
+# Can't customize vertex id when id strategy is 'AUTOMATIC': Firstly add vertex without id(id strategy is 'AUTOMATIC'), then write it into file, at last read file , call add vertex with id
```
Comment (javeme) [inline hugegraph-test/src/main/resources/fast-methods.filter]: remove space in "at last read file ,"

File: hugegraph-test/src/main/resources/methods.filter [hunk 15]
```diff
@@ -107,9 +106,13 @@ org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasTest.Traversals.g_
 org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasTest.Traversals.g_V_hasXage_withoutX27_29X_count: Unsupported relation 'NEQ'
 org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasTest.Traversals.g_V_hasXname_gtXmX_andXcontainingXoXXX: Unsupported predicate 'containing(o)'
 
-# Unsupport edge label 'created': 'software' -> 'person'
+# Unsupport edge label 'created': 'software' -> 'person', has existed an edgelabel (created: person -> software) in this case
 org.apache.tinkerpop.gremlin.process.traversal.step.map.AddEdgeTest.Traversals.g_V_hasXname_markoX_asXaX_outEXcreatedX_asXbX_inV_addEXselectXbX_labelX_toXaX: Unsupport edge from inV to outV
 
+# Can't customize vertex id when id strategy is 'AUTOMATIC': Firstly add vertex without id(id strategy is 'AUTOMATIC'), then write it into file, at last read file , call add vertex with id
```
Comment (javeme) [inline hugegraph-test/src/main/resources/methods.filter]: ditto

File: hugegraph-test/pom.xml (hunk 12 above)
Comment (Linary) [reply hugegraph-test/pom.xml]: travis doesn't have perf test

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/QueryHolder.java [hunk 16]
```diff
@@ -31,6 +31,24 @@
...
+        }
+
+        @Override
+        public boolean hasNext() {
+            return false;
+        }
+
+        @Override
+        public E next() {
+            return null;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/QueryHolder.java]: throw NoSuchElement

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/QueryHolder.java [hunk 17]
```diff
@@ -31,6 +31,24 @@
 
     public static final String SYSPROP_PAGE = "~page";
 
+    public static class PageElementIterator<E> implements Iterator<E>,
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/QueryHolder.java]: add const var EMPTY

File: pom.xml [hunk 18]
```diff
@@ -205,7 +205,7 @@
             <dependency>
                 <groupId>org.apache.tinkerpop</groupId>
                 <artifactId>gremlin-groovy-test</artifactId>
-                <version>${tinkerpop.version}</version>
+                <version>3.2.11</version>
```
Comment (javeme) [inline pom.xml]: keep ${tinkerpop.version}

Comment (Linary) [reply pom.xml]: gremlin-groovy-test the newest version is 3.2.11

File: hugegraph-test/pom.xml [hunk 19]
```diff
@@ -136,6 +133,8 @@
                     <execution>
                         <id>tinkerpop-process-test</id>
                         <configuration>
+                            <!-- ReadTest.Traversals.class need it -->
```
Comment (javeme) [inline hugegraph-test/pom.xml]: seems should remove tinkerpop-process-test
